fendermint/vm/interpreter/src/fvm/state/ipc.rs

```diff
@@ -263,16 +275,50 @@ impl<DB: Blockstore> GatewayCaller<DB> {
         &self,
         state: &mut FvmExecState<DB>,
         cross_messages: Vec<IpcEnvelope>,
-    ) -> anyhow::Result<FvmApplyRet> {
-        let messages = cross_messages
-            .into_iter()
-            .map(xnet_messaging_facet::IpcEnvelope::try_from)
-            .collect::<Result<Vec<_>, _>>()
-            .context("failed to convert cross messages")?;
-        let r = self
+        final_ret: &mut TopDownApplyRetAggregator,
+    ) -> anyhow::Result<()> {
+        for msg in cross_messages {
+            let msg = xnet_messaging_facet::IpcEnvelope::try_from(msg)
+                .context("failed to convert cross messages")?;
+            self.apply_cross_msg(state, msg, final_ret)?;
+        }
+
+        Ok(())
+    }
+
+    fn apply_cross_msg(
+        &self,
+        state: &mut FvmExecState<DB>,
+        msg: xnet_messaging_facet::IpcEnvelope,
+        final_ret: &mut TopDownApplyRetAggregator,
+    ) -> anyhow::Result<()> {
+        let (msg, success, ret) = match self
+            .xnet
+            .call_with_return(state, |c| c.apply_msg_with_ret(msg.clone()))
+        {
+            Ok(v) => {
+                let (has_return, apply_ret) = v.decoded_ret()?;
+
+                final_ret.agg_return(v.into_return());
+
+                if !has_return {
+                    return Ok(());
+                }
+
+                (apply_ret.cross_msg, apply_ret.success, apply_ret.ret)
+            }
+            Err(e) => {
+                let ret = ethers::abi::encode(&[Token::String(e.to_string())]);
```

Comment:
This `e` is the failure case of the `anyhow::Result`, right? I don't think you should use `e.to_string()` for sending back via ledger what happened to the contract. It won't be deterministic for many reasons:
* It's whatever `anyhow` decides to format the errors as
* Another non-Rust implementation, or Rust but not using this library, will have something completely different
* The concrete value returned by `call_with_return` (ie. `bail!("failed to execute contract call to {}:\ncode: {}\nerror: {:?}\ninfo: {}"`) was meant for a readable error in the logs, not as a programmatically accessible error. 

If you want to send back something, you should look at `try_call_with_ret` which returns a `CallError` if things go wrong, which contains the exit code and failure info, which should be deterministic. This is wrapped in `Ok(Err(...))` - if the outer `Ok` is an `Err` then the whole invocation failed and there's nothing to send back.

So, I reckon what you send back has to be similar to what you receive from the FVM: an exit code, a (potentially empty) returned bytes, and the separate failure info. Don't do anything to these if you want them to be deterministic.

Reply:
This is why it would be good if the host wasn't involved in the _ack_ handling: if it's all handled within Solidity then there is little that can be non-deterministic about it.